pkg/scheduler/framework/plugins/volumerestrictions/volume_restrictions.go

```diff
@@ -338,15 +340,110 @@ func (pl *VolumeRestrictions) EventsToRegister() []framework.ClusterEventWithHin
-		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}},
+		{Event: framework.ClusterEvent{Resource: framework.Node, ActionType: framework.Add}, QueueingHintFn: pl.isSchedulableAfterNodeAdded},
 		// Pods may fail to schedule because the PVC it uses has not yet been created.
 		// This PVC is required to exist to check its access modes.
-		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}},
+		{Event: framework.ClusterEvent{Resource: framework.PersistentVolumeClaim, ActionType: framework.Add | framework.Update}, QueueingHintFn: pl.isSchedulableAfterPersistentVolumeClaimChange},
 	}
 }
 
+// isSchedulableAfterPodDeleted is invoked whenever a pod deleted,
+// It checks whether the deleted pod will conflict with volumes of other pods on the same node
+func (pl *VolumeRestrictions) isSchedulableAfterPodDeleted(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	oldPod, _, err := util.As[*v1.Pod](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterPodDeleted", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	readWriteOncePVCUseByAnotherPod, err := pl.readWriteOncePodPVCUseByAnotherPod(context.TODO(), pod)
+	if err != nil {
+		logger.Error(err, "encountered an error when obtaining PVC according to the pod", "pod", klog.KObj(pod))
+	}
+
+	if pl.enableReadWriteOncePod && readWriteOncePVCUseByAnotherPod {
+		logger.V(4).Info("The Pod uses a PVC with ReadWriteOncePod access mode that is occupied by another Pod.", "pod", klog.KObj(pod))
+		return framework.QueueAfterBackoff
+	}
+
+	nodeInfo := framework.NewNodeInfo(oldPod)
+	if !satisfyVolumeConflicts(pod, nodeInfo) {
+		return framework.QueueAfterBackoff
+	}
+
+	return framework.QueueSkip
+}
+
+// isSchedulableAfterNodeAdded is invoked whenever a node added, It checks whether
+// that change made a previously unschedulable pod schedulable.
+func (pl *VolumeRestrictions) isSchedulableAfterNodeAdded(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) framework.QueueingHint {
+	_, newNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		logger.Error(err, "unexpected objects in isSchedulableAfterNodeAdded", "oldObj", oldObj, "newObj", newObj)
+		return framework.QueueAfterBackoff
+	}
+
+	readWriteOncePVCUseByAnotherPod, err := pl.readWriteOncePodPVCUseByAnotherPod(context.TODO(), pod)
+	if err != nil {
+		logger.Error(err, "encountered an error when obtaining PVC according to the pod", "pod", klog.KObj(pod))
+	}
+
+	if pl.enableReadWriteOncePod && readWriteOncePVCUseByAnotherPod {
+		logger.V(4).Info("The Pod uses a PVC with ReadWriteOncePod access mode that is occupied by another Pod.", "pod", klog.KObj(pod))
+		return framework.QueueAfterBackoff
+	}
+
+	logger.V(4).Info("node was added", "pod", klog.KObj(pod), "node", klog.KObj(newNode))
+	return framework.QueueSkip
```

Comment:
> IMO, the QueueingHintFn is used to do some pre calculation in the `Filter` logic of plugin. But we should always do `PreFilter` before as this only involves the pod. Which in `volume_restriction`, we check the RWOP pvcs of the pod.

Based on our discussion above, if we are checking just to return `QueueSkip`, why don't we just return `QueueSkip`?

Reply:
> If readWriteOncePVCUseByAnotherPod then we return skip, otherwise we return backoff, and we don't need to convert obj to node here.

Because I think if we don't have a conflict rwop pvc, and a new node added. We should try to reschedule it(return backoff)🤔️.  Let's say we didn't enable RWOP pvc.  This function will always hit line 387, can you expalin why we should always skip scheduling when a new node added?

If all the pod that conflicting with current pod are on node1, and a new node2 is added, and there's no new pod on node2. I think we can schedule current pod to node2(if it doesn't have a conflict rwop pvc).

Reply:
The new node won't have an impact on pod scheduling due to volume_restriction, won't it? So it should be skipped, maybe we should refer again to [discussion](https://github.com/kubernetes/kubernetes/pull/118551#discussion_r1269193782)

Reply:
> If all the pod that conflicting with current pod are on node1, and a new node2 is added, and there's no new pod on node2. I think we can schedule current pod to node2(if it doesn't have a conflict rwop pvc).

I think in this case , it will? We can always put pod to unschedulable pod pool in this case. If a new node is added and there's no pods on this node, we can definately schedule this pod to the new node🤔️

Reply:
Yes, so we return QueueSkip and leave it up to other events to decide whether it can be dispatched or not

ref:
> * if more than one QueueingHintFn returns QueueImmediately, it queues Pod to activeQ.
> * If no QueueingHintFn returns QueueImmediately and more than one plugin returns QueueAfterBackoff, it queues Pod to backoffQ if Pod is backing off, or to activeQ if Pod's backoff has already finished.
> * If all QueueingHintFn return QueueSkip, it puts this pod back to the unschedulable pod pool